File: ios/BluetoothGaming/Sources/BluetoothGaming/ProtocolCodec.swift
import Foundation

public enum LocXXBleUUIDs {
    public static let service = UUID(uuidString: "A0B40001-9267-4521-8F90-ACCD260FF718")!
    public static let clientToHost = UUID(uuidString: "A0B40002-9267-4521-8F90-ACCD260FF718")!
    public static let hostToClient = UUID(uuidString: "A0B40003-9267-4521-8F90-ACCD260FF718")!
}

public enum WireMessageType {
    public static let clientHello: UInt8 = 0x01
    public static let hostWelcome: UInt8 = 0x02
    public static let ack: UInt8 = 0x03
    public static let ping: UInt8 = 0x04
    public static let pong: UInt8 = 0x05
    public static let appPayload: UInt8 = 0x10
}

public let wireVersion: UInt8 = 0x01
public let maxFramePayload = 4096

public struct DecodedFrame: Equatable {
    public let wireVersion: UInt8
    public let messageType: UInt8
    public let payload: Data
}

public struct ClientHelloPayload: Equatable {
    public let protocolVersion: UInt8
    public let displayName: String
    public let clientNonce: Data
}

public struct HostWelcomePayload: Equatable {
    public let protocolVersion: UInt8
    public let playerId: UInt8
    public let sessionNonce: Data
}

public struct AckPayload: Equatable {
    public let acknowledgedMessageType: UInt8
    public let correlationId: UInt32
}

public enum ProtocolCodec {
    public static func encodeFrame(messageType: UInt8, payload: Data = Data()) -> Data {
        precondition(payload.count <= maxFramePayload)
        var buf = Data()
        buf.append(wireVersion)
        buf.append(messageType)
        var len = UInt16(payload.count).bigEndian
        buf.append(Data(bytes: &len, count: 2))
        buf.append(payload)
        return buf
    }

    public static func decodeFrame(_ frame: Data) throws -> DecodedFrame {
        guard frame.count >= 4 else { throw NSError(domain: "ProtocolCodec", code: 1) }
        let ver = frame[0]
        guard ver == wireVersion else { throw NSError(domain: "ProtocolCodec", code: 2) }
        let type = frame[1]
        let len = Int(frame[2]) << 8 | Int(frame[3])
        guard len <= maxFramePayload, frame.count == 4 + len else { throw NSError(domain: "ProtocolCodec", code: 3) }
        let payload = frame.subdata(in: 4 ..< (4 + len))
        return DecodedFrame(wireVersion: ver, messageType: type, payload: payload)
    }

    public static func buildClientHello(protocolVersion: UInt8, displayName: String, clientNonce: Data) throws -> Data {
        guard clientNonce.count == 16 else { throw NSError(domain: "ProtocolCodec", code: 4) }
        let nameData = Data(displayName.utf8)
        guard nameData.count <= 64 else { throw NSError(domain: "ProtocolCodec", code: 5) }
        var p = Data()
        p.append(protocolVersion)
        p.append(UInt8(nameData.count))
        p.append(nameData)
        p.append(clientNonce)
        return encodeFrame(messageType: WireMessageType.clientHello, payload: p)
    }

    public static func parseClientHello(_ payload: Data) throws -> ClientHelloPayload {
        guard payload.count >= 2 + 16 else { throw NSError(domain: "ProtocolCodec", code: 6) }
        let pv = payload[0]
        let nl = Int(payload[1])
        guard 2 + nl + 16 == payload.count else { throw NSError(domain: "ProtocolCodec", code: 7) }
        let name = String(data: payload.subdata(in: 2 ..< (2 + nl)), encoding: .utf8) ?? ""
        let nonce = payload.subdata(in: (2 + nl) ..< (2 + nl + 16))
        return ClientHelloPayload(protocolVersion: pv, displayName: name, clientNonce: nonce)
    }

    public static func buildHostWelcome(protocolVersion: UInt8, playerId: UInt8, sessionNonce: Data) throws -> Data {
        guard sessionNonce.count == 16 else { throw NSError(domain: "ProtocolCodec", code: 8) }
        var p = Data()
        p.append(protocolVersion)
        p.append(playerId)
        p.append(sessionNonce)
        return encodeFrame(messageType: WireMessageType.hostWelcome, payload: p)
    }

    public static func parseHostWelcome(_ payload: Data) throws -> HostWelcomePayload {
        guard payload.count == 18 else { throw NSError(domain: "ProtocolCodec", code: 9) }
        let nonce = payload.subdata(in: 2 ..< 18)
        return HostWelcomePayload(protocolVersion: payload[0], playerId: payload[1], sessionNonce: nonce)
    }

    public static func buildAck(acknowledgedMessageType: UInt8, correlationId: UInt32) -> Data {
        var p = Data()
        p.append(acknowledgedMessageType)
        var cid = correlationId.bigEndian
        p.append(Data(bytes: &cid, count: 4))
        return encodeFrame(messageType: WireMessageType.ack, payload: p)
    }

    public static func parseAck(_ payload: Data) throws -> AckPayload {
        guard payload.count == 5 else { throw NSError(domain: "ProtocolCodec", code: 10) }
        let mt = payload[0]
        let cid = (UInt32(payload[1]) << 24) | (UInt32(payload[2]) << 16) | (UInt32(payload[3]) << 8) | UInt32(payload[4])
        return AckPayload(acknowledgedMessageType: mt, correlationId: cid)
    }
}

File: android/lan-gaming/src/main/kotlin/com/locxx/langaming/ProtocolCodec.kt
package com.locxx.langaming

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID

object LocXXBleUuids {
    val SERVICE: UUID = UUID.fromString("A0B40001-9267-4521-8F90-ACCD260FF718")
    val CLIENT_TO_HOST: UUID = UUID.fromString("A0B40002-9267-4521-8F90-ACCD260FF718")
    val HOST_TO_CLIENT: UUID = UUID.fromString("A0B40003-9267-4521-8F90-ACCD260FF718")
}

object WireMessageType {
    const val CLIENT_HELLO: Byte = 0x01
    const val HOST_WELCOME: Byte = 0x02
    const val ACK: Byte = 0x03
    const val PING: Byte = 0x04
    const val PONG: Byte = 0x05
    const val APP_PAYLOAD: Byte = 0x10
}

const val WIRE_VERSION: Byte = 0x01
const val MAX_FRAME_PAYLOAD = 4096

const val LOCXX_LAN_PORT = 28765

object ProtocolCodec {

    fun encodeFrame(messageType: Byte, payload: ByteArray = ByteArray(0)): ByteArray {
        require(payload.size <= MAX_FRAME_PAYLOAD) { "payload too large" }
        return ByteBuffer.allocate(4 + payload.size).order(ByteOrder.BIG_ENDIAN).apply {
            put(WIRE_VERSION)
            put(messageType)
            putShort(payload.size.toShort())
            put(payload)
        }.array()
    }

    fun decodeFrame(frame: ByteArray): Result<DecodedFrame> {
        if (frame.size < 4) return Result.failure(IllegalArgumentException("frame too short"))
        val buf = ByteBuffer.wrap(frame).order(ByteOrder.BIG_ENDIAN)
        val ver = buf.get()
        if (ver != WIRE_VERSION) return Result.failure(IllegalArgumentException("bad wire version"))
        val type = buf.get()
        val len = buf.short.toInt() and 0xFFFF
        if (len > MAX_FRAME_PAYLOAD) return Result.failure(IllegalArgumentException("bad length"))
        if (frame.size != 4 + len) return Result.failure(IllegalArgumentException("size mismatch"))
        val payload = ByteArray(len)
        buf.get(payload)
        return Result.success(DecodedFrame(ver, type, payload))
    }

    fun buildClientHello(protocolVersion: Byte, displayName: String, clientNonce: ByteArray): ByteArray {
        require(clientNonce.size == 16)
        val nameBytes = displayName.toByteArray(Charsets.UTF_8)
        require(nameBytes.size <= 64)
        val payload = ByteBuffer.allocate(1 + 1 + nameBytes.size + 16).apply {
            put(protocolVersion)
            put(nameBytes.size.toByte())
            put(nameBytes)
            put(clientNonce)
        }.array()
        return encodeFrame(WireMessageType.CLIENT_HELLO, payload)
    }

    fun parseClientHello(payload: ByteArray): Result<ClientHelloPayload> {
        if (payload.size < 1 + 1 + 16) return Result.failure(IllegalArgumentException("hello too short"))
        val pv = payload[0]
        val nl = payload[1].toInt() and 0xFF
        if (2 + nl + 16 != payload.size) return Result.failure(IllegalArgumentException("hello name"))
        val name = String(payload, 2, nl, Charsets.UTF_8)
        val nonce = payload.copyOfRange(2 + nl, 2 + nl + 16)
        return Result.success(ClientHelloPayload(pv, name, nonce))
    }

    fun buildHostWelcome(protocolVersion: Byte, playerId: Byte, sessionNonce: ByteArray): ByteArray {
        require(sessionNonce.size == 16)
        val payload = ByteBuffer.allocate(1 + 1 + 16).apply {
            put(protocolVersion)
            put(playerId)
            put(sessionNonce)
        }.array()
        return encodeFrame(WireMessageType.HOST_WELCOME, payload)
    }

    fun parseHostWelcome(payload: ByteArray): Result<HostWelcomePayload> {
        if (payload.size != 18) return Result.failure(IllegalArgumentException("welcome size"))
        return Result.success(
            HostWelcomePayload(
                protocolVersion = payload[0],
                playerId = payload[1],
                sessionNonce = payload.copyOfRange(2, 18)
            )
        )
    }

    fun buildAck(acknowledgedMessageType: Byte, correlationId: Int): ByteArray {
        val p = ByteBuffer.allocate(5).order(ByteOrder.BIG_ENDIAN).apply {
            put(acknowledgedMessageType)
            putInt(correlationId)
        }.array()
        return encodeFrame(WireMessageType.ACK, p)
    }

    fun parseAck(payload: ByteArray): Result<AckPayload> {
        if (payload.size != 5) return Result.failure(IllegalArgumentException("ack size"))
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
        val mt = buf.get()
        val cid = buf.int
        return Result.success(AckPayload(mt, cid))
    }
}

data class DecodedFrame(
    val wireVersion: Byte,
    val messageType: Byte,
    val payload: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as DecodedFrame
        if (wireVersion != other.wireVersion) return false
        if (messageType != other.messageType) return false
        if (!payload.contentEquals(other.payload)) return false
        return true
    }

    override fun hashCode(): Int {
        var result = wireVersion.hashCode()
        result = 31 * result + messageType.hashCode()
        result = 31 * result + payload.contentHashCode()
        return result
    }
}

data class ClientHelloPayload(
    val protocolVersion: Byte,
    val displayName: String,
    val clientNonce: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as ClientHelloPayload
        if (protocolVersion != other.protocolVersion) return false
        if (displayName != other.displayName) return false
        if (!clientNonce.contentEquals(other.clientNonce)) return false
        return true
    }

    override fun hashCode(): Int {
        var result = protocolVersion.hashCode()
        result = 31 * result + displayName.hashCode()
        result = 31 * result + clientNonce.contentHashCode()
        return result
    }
}

data class HostWelcomePayload(
    val protocolVersion: Byte,
    val playerId: Byte,
    val sessionNonce: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as HostWelcomePayload
        if (protocolVersion != other.protocolVersion) return false
        if (playerId != other.playerId) return false
        if (!sessionNonce.contentEquals(other.sessionNonce)) return false
        return true
    }

    override fun hashCode(): Int {
        var result = protocolVersion.hashCode()
        result = 31 * result + playerId.hashCode()
        result = 31 * result + sessionNonce.contentHashCode()
        return result
    }
}

data class AckPayload(val acknowledgedMessageType: Byte, val correlationId: Int)
